# Tarea 1: Deep Learning 2025
- Alumno: Francisco Morales Arias
- Profesor: Christian Aguilera
  
Desarrollar un análisis exploratorio de datos (EDA) y modelación para el dataset asignado.

---

## 1 Análisis exploratorio
detección de valores nulos, duplicados, errores, outliers y descripción estadística general del dataset (con visualizaciones).

### Desarrollo


In [1]:
import pandas as pd
import numpy as np
from sklearn.tree import DecisionTreeClassifier, plot_tree
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, precision_score, recall_score
from sklearn.preprocessing import OneHotEncoder
from sklearn.impute import SimpleImputer
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler

import warnings
warnings.filterwarnings('ignore')

file = pd.read_csv('factory_sensors_draw2.csv')

In [2]:

file.tail(), file.columns, file.size

(            s1        s2         s3        s4        s5        s6         s7  \
 6235       NaN -5.547598   2.786233 -1.064707  3.345072 -8.343020   5.339206   
 6236       NaN  2.061614 -10.405416 -1.944145 -5.399278       NaN  11.396246   
 6237  1.797639  7.720596   7.791849  4.624720  1.263150 -3.828952 -15.489657   
 6238  8.039895       NaN   8.389923 -6.171220 -1.512421  0.795013   6.282524   
 6239  0.431476  4.516479  -2.091449 -4.991664 -2.198562 -5.752324  -1.665545   
 
             s8         s9       s10  ...     temp_c  pressure_bar  \
 6235  9.253984 -12.359032  4.787808  ...  26.048537      4.493444   
 6236 -3.799092   9.744953  1.288012  ...  52.881378      4.729492   
 6237       NaN        NaN  3.874763  ...  42.399639      5.767572   
 6238 -3.872433   1.853285  6.755307  ...        NaN      4.259805   
 6239       NaN  14.884312 -2.022313  ...  49.767743      4.178095   
 
       humidity_pct  machine_type    shift  plant_region  operator_id  line_id  \
 6235   

In [3]:
copy_df = file.copy()

In [4]:
describe_df = copy_df.describe()
describe_df

s1           s2           s3           s4           s5  \
count  5569.000000  5744.000000  5740.000000  5567.000000  5739.000000   
mean     -0.046408    -0.359728     0.151861     0.228111    -0.083065   
std       9.689571    11.753444    12.788576    13.870738     9.531714   
min    -154.885812  -234.324678  -146.267270  -287.393007  -159.965100   
25%      -3.817762    -4.158819    -4.384155    -4.647781    -3.462575   
50%       0.056846    -0.288866     0.061808     0.207675     0.133785   
75%       3.637580     4.039275     4.513516     5.002652     3.534238   
max     149.472209   190.331128   187.437937   175.517693   140.908923   

                s6           s7           s8           s9          s10  ...  \
count  5576.000000  5739.000000  5580.000000  5557.000000  5562.000000  ...   
mean     -0.077371     0.081703    -0.156282    -0.289547     0.013984  ...   
std      10.630547    12.449184    11.502267    13.371879     9.575010  ...   
min    -229.826662  -181.778493  -151.543729  -185.628837  -154.510493  ...   
25%      -3.717650    -4.234571    -4.475151    -4.806461    -3.532811  ...   
50%       0.070994    -0.121512    -0.108053    -0.162190    -0.007889  ...   
75%       3.891414     3.977762     4.417187     4.467643     3.448989  ...   
max     140.023327   206.945579   224.629955   215.932581   144.284203  ...   

               s18          s19          s20  vibration_rms       temp_c  \
count  5570.000000  5568.000000  5734.000000    5563.000000  5573.000000   
mean     -0.148162     0.008992    -0.052697       2.359610    46.237942   
std      12.999301    13.196476     9.736059       3.616105    56.884372   
min    -202.912857  -188.800229  -130.107977     -46.028533  -632.167223   
25%      -4.430903    -4.554918    -3.580737       1.247840    35.232839   
50%      -0.137658     0.169119    -0.072631       2.037745    39.939445   
75%       4.403312     5.010587     3.375525       2.839293    44.910286   
max     201.146002   211.357384   152.271345      66.408060   725.215265   

       pressure_bar  humidity_pct  operator_id      line_id  quality_score  
count   5573.000000   5739.000000  5569.000000  5571.000000    5746.000000  
mean       5.613677     62.054918   124.448734     2.995692      86.929465  
std        7.185770     80.743225    14.455538     1.428229     104.048786  
min      -99.658064   -976.342793   100.000000     1.000000    -964.369440  
25%        4.219318     44.647397   112.000000     2.000000      59.581931  
50%        5.051424     55.154022   124.000000     3.000000      72.721974  
75%        5.898466     66.207613   137.000000     4.000000      90.394332  
max       98.457929   1070.969061   149.000000     5.000000    1942.546474  

[8 rows x 27 columns]

In [5]:
def cv(df):
    cv = df.loc['std'] / df.loc['mean']
    cv = cv.replace([np.inf, -np.inf], np.nan)
    
    print("\n=== Coeficiente de Variación (CV) ===")
    print(np.absolute(cv).sort_values(ascending=False))
    
    print("\n=== CV en Porcentaje (%) ===")
    print(np.absolute((cv * 100).round(2)).sort_values(ascending=False))
    
    return

In [6]:
cv(describe_df)


=== Coeficiente de Variación (CV) ===
s19              1467.583648
s10               684.706939
s11               377.924688
s15               355.536803
s1                208.789398
s20               184.755815
s7                152.370837
s6                137.396234
s13               122.664177
s5                114.750630
s16               111.562103
s12               100.458705
s18                87.736804
s3                 84.212285
s8                 73.599358
s14                73.172610
s4                 60.806933
s9                 46.182047
s17                43.705678
s2                 32.673133
vibration_rms       1.532501
humidity_pct        1.301158
pressure_bar        1.280047
temp_c              1.230253
quality_score       1.196933
line_id             0.476761
operator_id         0.116157
dtype: float64

=== CV en Porcentaje (%) ===
s19              146758.36
s10               68470.69
s11               37792.47
s15               35553.68
s1                20878.9

In [7]:
copy_df.isnull().sum().sort_values(ascending=False)

shift            779
s9               683
s10              678
vibration_rms    677
s16              676
s12              676
plant_region     674
s11              673
s4               673
s19              672
operator_id      671
s1               671
s14              671
s18              670
line_id          669
machine_type     668
s15              668
s13              668
pressure_bar     667
temp_c           667
timestamp        665
s6               664
s8               660
s20              506
humidity_pct     501
s5               501
s7               501
s3               500
s2               496
s17              495
quality_score    494
dtype: int64

In [8]:
copy_df.duplicated().sum()

np.int64(228)

In [9]:
df_T= copy_df.T

In [10]:
df_T.isnull().sum().sort_values(ascending=False)

3086    11
3462    11
3020    10
2631    10
6080    10
        ..
3845     0
3900     0
3907     0
3910     0
54       0
Length: 6240, dtype: int64

In [11]:
df_T.duplicated().sum()

np.int64(0)

### Comentarios
Se observa que el tamaño del dataset el cual corresponde a un total de 6240 filas y 31 columnas o predictores, de entre estos, los datos que se encuentran corresponden a datos entregados por sensores (predictores 1 al 20), parámetros como vibraciones del sistema, temperatura en celsius, presión, húmedad, y datos cuantitativos como el tipo de máquina, el turno, la ubicación de la planta, indentificadores, hora y calificación de calidad. Por lo tanto, es posible que este dataset sea parte de un informe que detalle el estado de un sistema de manera periódica, esto con el objetivo de realizar algún tipo de mantenimiento predictivo.
Respecto a los datos del data frame, el análisis exploratorio arrojó los siguientes datos:
- Los datos se encuentran desescalados.
- Existen una enorme cantidad de valores nulos, algunas columnas rondan los 800 datos faltantes.
- Los datos se encuentran increiblemente dispersos, en los resultados del coeficiente de variación el valor del sensor 19 alcanza una variación del 146758%.
- Hay un total de 228 filas y 0 columnas repetidas

Uno de los primeros detalles que salta a la vista, es como la columna "quality_score" presenta valores NaN. Si se utiliza esta columna como la salida del módelo, entonces los datos de las filas sin calificación deberían de ser descartados. Esto se llevará a cabo en la siguiente sección.

---

## 2. Preprocesamiento

Transformación a datos numéricos, eliminación de columnas redundantes, imputación de valores faltantes y técnicas de balanceo de clases si es necesario.

### Eliminación primaria

Tal y como se mencionó en la sección anterior, la columna "quality_score" contiene elementos nulos, por lo tanto ese tipo de filas no presenta ningún tipo de aporte para el entrenamiento. El descarte de estos elementos es de un volumen el cual no es menor, sin embargo, se considera que a pesar de ser una cantidad considerable de datos, al no presentar utilidad se pueden despreciar fácilmente.


In [12]:
df_y_true = copy_df.dropna(subset=['quality_score'])
df_y_true['quality_score']

0        51.345036
2       103.537986
3        48.850995
4        70.346204
5        56.059338
           ...    
6234     68.464070
6235     80.836287
6236     54.734622
6237     86.463000
6239     73.100548
Name: quality_score, Length: 5746, dtype: float64

### Tiempo

In [13]:
df_y_true['timestamp']

0       2024-04-29 14:57:00
2       2024-04-30 04:28:00
3                       NaN
4       2024-01-28 04:43:00
5       2024-06-01 14:14:00
               ...         
6234    2024-06-12 14:09:00
6235    2024-01-31 08:41:00
6236                    NaN
6237    2024-05-18 06:13:00
6239    2024-01-13 08:12:00
Name: timestamp, Length: 5746, dtype: object

In [14]:
df_y_true = df_y_true.copy()
df_y_true['timestamp'] = pd.to_datetime(df_y_true['timestamp'], format='mixed', errors='coerce')
print(f"Fechas inválidas encontradas: {df_y_true['timestamp'].isnull().sum()}")
porcentaje_malo = df_y_true['timestamp'].isnull().sum() / len(df_y_true)

print(f"Porcentaje de filas con fechas erróneas: {porcentaje_malo * 100:.2f}%")
print(f"Tipo de dato de la columna 'timestamp': {df_y_true['timestamp'].dtype}")

Fechas inválidas encontradas: 667
Porcentaje de filas con fechas erróneas: 11.61%
Tipo de dato de la columna 'timestamp': datetime64[ns]


In [15]:
df_y_true['timestamp']

0      2024-04-29 14:57:00
2      2024-04-30 04:28:00
3                      NaT
4      2024-01-28 04:43:00
5      2024-06-01 14:14:00
               ...        
6234   2024-06-12 14:09:00
6235   2024-01-31 08:41:00
6236                   NaT
6237   2024-05-18 06:13:00
6239   2024-01-13 08:12:00
Name: timestamp, Length: 5746, dtype: datetime64[ns]

In [16]:
df_y_true['hora'] = df_y_true['timestamp'].dt.hour
df_y_true['mes'] = df_y_true['timestamp'].dt.month
df_y_true['dia_semana'] = df_y_true['timestamp'].dt.dayofweek

df_v2 = df_y_true.drop('timestamp',axis=1)

len(df_v2.columns),df_v2.columns

(33,
 Index(['s1', 's2', 's3', 's4', 's5', 's6', 's7', 's8', 's9', 's10', 's11',
        's12', 's13', 's14', 's15', 's16', 's17', 's18', 's19', 's20',
        'vibration_rms', 'temp_c', 'pressure_bar', 'humidity_pct',
        'machine_type', 'shift', 'plant_region', 'operator_id', 'line_id',
        'quality_score', 'hora', 'mes', 'dia_semana'],
       dtype='object'))

### Tratamiento de NaNs: Eliminación de filas
Considerando la enorme cantidad de datos faltantes, lo primero que se hará, será eliminar filas que tengan mas de 7 elementos faltantes. De esta manera se crea una reducción del dataframe, la cual genera una reducción de los datos menor de un 5%.

In [17]:
df_v2.T.isnull().sum().sort_values()

4560     0
2283     0
2287     0
2254     0
1197     0
        ..
3735    11
3086    11
5939    11
1261    11
3492    11
Length: 5746, dtype: int64

In [18]:
df_v3 = df_v2.dropna(thresh=len(df_v2.columns)-7)
print("\n=== porcentaje de filas restantes ===")
print((len(df_v3)/len(df_y_true) * 100))
print("filas restantes:", len(df_v3), "\nfilas originales:", len(df_y_true))


=== porcentaje de filas restantes ===
97.63313609467455
filas restantes: 5610 
filas originales: 5746


### Estandarización de datos
Existen una serie de datos cuantitativos los cuales es pertinente convertir a valores numéricos, para esto se hace uso de la librería sklearn.

In [19]:
X = df_v3.drop('quality_score', axis=1)
y = df_v3['quality_score']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [20]:
categorical_columns = X_train.select_dtypes(include=['object']).columns.tolist()
encoder = OneHotEncoder(sparse_output=False,handle_unknown='ignore')
one_hot_en_train = encoder.fit_transform(X_train[categorical_columns])
one_hot_en_test = encoder.transform(X_test[categorical_columns])

one_hot_df_train = pd.DataFrame(one_hot_en_train,
                                columns=encoder.get_feature_names_out(categorical_columns), 
                                index=X_train.index)

one_hot_df_test = pd.DataFrame(one_hot_en_test, 
                               columns=encoder.get_feature_names_out(categorical_columns), 
                               index=X_test.index)

X_train_v2 = pd.concat([X_train, one_hot_df_train], axis=1)
X_test_v2 = pd.concat([X_test, one_hot_df_test], axis=1)

X_train_v2 = X_train_v2.drop(categorical_columns, axis=1)
X_test_v2 = X_test_v2.drop(categorical_columns, axis=1)
len(X_train_v2.columns),X_train_v2.columns,len(X_test_v2.columns),X_test_v2.columns

(41,
 Index(['s1', 's2', 's3', 's4', 's5', 's6', 's7', 's8', 's9', 's10', 's11',
        's12', 's13', 's14', 's15', 's16', 's17', 's18', 's19', 's20',
        'vibration_rms', 'temp_c', 'pressure_bar', 'humidity_pct',
        'operator_id', 'line_id', 'hora', 'mes', 'dia_semana', 'machine_type_A',
        'machine_type_B', 'machine_type_C', 'machine_type_nan', 'shift_day',
        'shift_evening', 'shift_night', 'shift_nan', 'plant_region_Centro',
        'plant_region_Norte', 'plant_region_Sur', 'plant_region_nan'],
       dtype='object'),
 41,
 Index(['s1', 's2', 's3', 's4', 's5', 's6', 's7', 's8', 's9', 's10', 's11',
        's12', 's13', 's14', 's15', 's16', 's17', 's18', 's19', 's20',
        'vibration_rms', 'temp_c', 'pressure_bar', 'humidity_pct',
        'operator_id', 'line_id', 'hora', 'mes', 'dia_semana', 'machine_type_A',
        'machine_type_B', 'machine_type_C', 'machine_type_nan', 'shift_day',
        'shift_evening', 'shift_night', 'shift_nan', 'plant_region_Centro',

In [21]:
try:
    X_train_v2.pop('machine_type_nan')
    X_train_v2.pop('shift_nan')
    X_train_v2.pop('plant_region_nan')
    X_test_v2.pop('machine_type_nan')
    X_test_v2.pop('shift_nan')
    X_test_v2.pop('plant_region_nan')
except:
    pass
len(X_train_v2.columns),X_train_v2.columns,len(X_test_v2.columns),X_test_v2.columns

(38,
 Index(['s1', 's2', 's3', 's4', 's5', 's6', 's7', 's8', 's9', 's10', 's11',
        's12', 's13', 's14', 's15', 's16', 's17', 's18', 's19', 's20',
        'vibration_rms', 'temp_c', 'pressure_bar', 'humidity_pct',
        'operator_id', 'line_id', 'hora', 'mes', 'dia_semana', 'machine_type_A',
        'machine_type_B', 'machine_type_C', 'shift_day', 'shift_evening',
        'shift_night', 'plant_region_Centro', 'plant_region_Norte',
        'plant_region_Sur'],
       dtype='object'),
 38,
 Index(['s1', 's2', 's3', 's4', 's5', 's6', 's7', 's8', 's9', 's10', 's11',
        's12', 's13', 's14', 's15', 's16', 's17', 's18', 's19', 's20',
        'vibration_rms', 'temp_c', 'pressure_bar', 'humidity_pct',
        'operator_id', 'line_id', 'hora', 'mes', 'dia_semana', 'machine_type_A',
        'machine_type_B', 'machine_type_C', 'shift_day', 'shift_evening',
        'shift_night', 'plant_region_Centro', 'plant_region_Norte',
        'plant_region_Sur'],
       dtype='object'))

### Tratamiento de NaNs: inputación de datos
A partir de este punto, el dataframe se encuentra principalmente limpio, y se procederá a hacer la separación entre los predictores X y el resultado Y.

In [22]:
imputer = SimpleImputer(missing_values=np.nan, strategy='mean')

array_train_imputado = imputer.fit_transform(X_train_v2)
array_test_imputado = imputer.transform(X_test_v2)


X_train_v3 = pd.DataFrame(array_train_imputado, 
                       columns=X_train_v2.columns, 
                       index=X_train_v2.index)

X_test_v3 = pd.DataFrame(array_test_imputado, 
                      columns=X_test_v2.columns, 
                      index=X_test_v2.index)

X_train_v3.isnull().sum(),X_test_v3.isnull().sum()

(s1                     0
 s2                     0
 s3                     0
 s4                     0
 s5                     0
 s6                     0
 s7                     0
 s8                     0
 s9                     0
 s10                    0
 s11                    0
 s12                    0
 s13                    0
 s14                    0
 s15                    0
 s16                    0
 s17                    0
 s18                    0
 s19                    0
 s20                    0
 vibration_rms          0
 temp_c                 0
 pressure_bar           0
 humidity_pct           0
 operator_id            0
 line_id                0
 hora                   0
 mes                    0
 dia_semana             0
 machine_type_A         0
 machine_type_B         0
 machine_type_C         0
 shift_day              0
 shift_evening          0
 shift_night            0
 plant_region_Centro    0
 plant_region_Norte     0
 plant_region_Sur       0
 dtype: int6

### Tratamiento de Outliers

In [23]:
def how_many_outliers(df):
    Q1 = df.quantile(0.25)
    Q3 = df.quantile(0.75)
    IQR = Q3 - Q1

    outliers = ((df < (Q1 - 1.5 * IQR)) | (df > (Q3 + 1.5 * IQR)))
    outliers= outliers.sum(axis=1).sort_values()
    outliers = outliers[outliers!=0]
    return outliers
    
outliers = how_many_outliers(X_train_v3)
outliers.sort_values(ascending=False)

3599    22
434     22
733     22
4722    22
5999    22
        ..
5561     1
367      1
2233     1
661      1
2676     1
Length: 3189, dtype: int64

In [24]:
numerical_cols = ['s1', 's2', 's3', 's4', 's5', 's6', 's7', 's8', 's9', 's10', 's11',
       's12', 's13', 's14', 's15', 's16', 's17', 's18', 's19', 's20',
       'vibration_rms', 'temp_c', 'pressure_bar', 'humidity_pct',
       'operator_id', 'line_id', 'hora', 'mes', 'dia_semana']

for col in numerical_cols:

    Q1 = X_train_v3[col].quantile(0.25)
    Q3 = X_train_v3[col].quantile(0.75)
    IQR = Q3 - Q1
    
    lower_fence = Q1 - (1.5 * IQR)
    upper_fence = Q3 + (1.5 * IQR)
    
    X_train_v3[col] = X_train_v3[col].clip(lower_fence, upper_fence)

    X_test_v3[col] = X_test_v3[col].clip(lower_fence, upper_fence)
    
how_many_outliers(X_train_v3).sort_values(ascending=False),how_many_outliers(X_test_v3).sort_values(ascending=False)

(2893    3
 2501    3
 2612    3
 3576    3
 4448    3
        ..
 5651    1
 4948    1
 605     1
 3080    1
 524     1
 Length: 2609, dtype: int64,
 4173    14
 4454    14
 5027    13
 3720    13
 2531    12
         ..
 4022     1
 5157     1
 4370     1
 3085     1
 6076     1
 Length: 757, dtype: int64)

In [25]:
X_test_v3.describe()

s1           s2           s3           s4           s5  \
count  1122.000000  1122.000000  1122.000000  1122.000000  1122.000000   
mean      0.046418    -0.203978     0.074034     0.177809     0.147721   
std       5.291492     5.938229     6.688228     6.618432     5.006751   
min     -13.270921   -14.843155   -15.649578   -16.613447   -12.493412   
25%      -3.163982    -3.734759    -4.510043    -4.034795    -3.124182   
50%      -0.045341    -0.319739     0.073880     0.205387    -0.087478   
75%       3.083908     3.593861     4.735761     3.920323     3.029883   
max      13.038960    14.952691    15.733632    16.966238    12.535011   

                s6           s7           s8           s9          s10  ...  \
count  1122.000000  1122.000000  1122.000000  1122.000000  1122.000000  ...   
mean     -0.180767    -0.145742    -0.166059    -0.470372    -0.153066  ...   
std       5.332969     6.054885     6.164656     6.559311     4.937837  ...   
min     -12.907631   -15.007789   -15.254333   -15.938144   -12.374268  ...   
25%      -3.576985    -3.722964    -4.069590    -4.376990    -3.425415  ...   
50%      -0.048546     0.278924    -0.123575    -0.020671     0.025205  ...   
75%       3.062203     3.399541     3.677827     3.396403     2.843384  ...   
max      13.022510    14.992013    15.343773    15.722006    12.214400  ...   

        dia_semana  machine_type_A  machine_type_B  machine_type_C  \
count  1122.000000     1122.000000     1122.000000     1122.000000   
mean      2.940355        0.350267        0.355615        0.178253   
std       1.922221        0.477266        0.478913        0.382896   
min       0.000000        0.000000        0.000000        0.000000   
25%       1.000000        0.000000        0.000000        0.000000   
50%       3.000000        0.000000        0.000000        0.000000   
75%       5.000000        1.000000        1.000000        0.000000   
max       6.000000        1.000000        1.000000        1.000000   

         shift_day  shift_evening  shift_night  plant_region_Centro  \
count  1122.000000     1122.00000  1122.000000          1122.000000   
mean      0.391266        0.29590     0.188948             0.537433   
std       0.488251        0.45665     0.391642             0.498819   
min       0.000000        0.00000     0.000000             0.000000   
25%       0.000000        0.00000     0.000000             0.000000   
50%       0.000000        0.00000     0.000000             1.000000   
75%       1.000000        1.00000     0.000000             1.000000   
max       1.000000        1.00000     1.000000             1.000000   

       plant_region_Norte  plant_region_Sur  
count         1122.000000       1122.000000  
mean             0.183601          0.180036  
std              0.387331          0.384388  
min              0.000000          0.000000  
25%              0.000000          0.000000  
50%              0.000000          0.000000  
75%              0.000000          0.000000  
max              1.000000          1.000000  

[8 rows x 38 columns]

In [26]:
stats = X_train_v3.describe()
cv(stats)


=== Coeficiente de Variación (CV) ===
s7                     1407.825285
s8                      643.126660
s5                      240.150151
s14                     209.494727
s9                      170.481617
s6                      151.857512
s12                     126.575678
s10                     122.931291
s1                      105.597089
s15                      94.607686
s18                      85.524848
s19                      82.470379
s11                      75.794510
s2                       70.961553
s13                      68.526788
s3                       62.225340
s16                      53.495348
s20                      49.701069
s4                       35.019504
s17                      22.622025
plant_region_Sur          2.206298
machine_type_C            2.112072
shift_night               2.099574
plant_region_Norte        2.082660
shift_evening             1.483790
machine_type_A            1.354157
machine_type_B            1.324936
shift_day       

### Análisis de correlación


In [27]:
corr_matrix = X_train_v3.corr().abs()

upper_triangle_mask = np.triu(np.ones(corr_matrix.shape), k=1).astype(bool)

upper_triangle_matrix = corr_matrix.where(upper_triangle_mask)

high_corr_series = upper_triangle_matrix.stack()

high_corr_pairs = high_corr_series[high_corr_series > 0.9]

if high_corr_pairs.empty:
    print("No se encontraron features fuertemente correlacionadas.")
else:
    print("Pares de features fuertemente correlacionadas (redundantes):")
    print(high_corr_pairs)

No se encontraron features fuertemente correlacionadas.


### Escalado

In [28]:
scaler = StandardScaler()

X_train_v3[numerical_cols] = scaler.fit_transform(X_train_v3[numerical_cols])

X_test_v3[numerical_cols] = scaler.transform(X_test_v3[numerical_cols])

X_train_clean = X_train_v3
X_test_clean = X_test_v3

In [29]:
X_test_clean

s1        s2        s3        s4        s5        s6        s7  \
1152 -1.091694 -1.023601  0.078341 -1.178557 -0.714932  0.208308 -0.833666   
4473 -0.350632 -0.642932 -0.209836  2.248952  0.950931  0.322664 -1.143051   
3679 -0.019636  0.602560  0.830026 -1.114036 -0.021635  1.079800  1.529033   
3186  1.551121 -0.968795 -0.900368 -0.262666 -0.177291 -0.015680  0.707325   
3826  0.322456 -0.087214  0.467026  0.858181 -0.632906  0.567837 -0.446744   
...        ...       ...       ...       ...       ...       ...       ...   
599   0.104958 -0.039246 -2.183498 -1.195261 -0.174989 -0.454281  0.152496   
3531  0.193166  0.476283  0.082825  0.001458  0.614113  2.433031 -1.014974   
3944  1.063749  0.544390 -0.161905 -1.132610 -1.202249 -1.134468 -0.823451   
1836  0.970517 -0.861223 -1.643296 -1.462789 -0.825016  0.254087  2.319947   
4090  0.444236 -0.039246 -0.395875  1.014881 -0.452281 -0.797299 -1.692809   

            s8        s9       s10  ...    dia_semana  machine_type_A  \
1152  0.398243  0.680514  0.013273  ... -5.631988e-03             0.0   
4473 -0.217703  0.922259 -0.782699  ...  1.049514e+00             1.0   
3679 -0.529611  1.493562 -0.442611  ... -2.342896e-16             1.0   
3186 -1.702287 -1.588008  0.013273  ...  5.219412e-01             1.0   
3826  0.248825 -0.583516 -0.292907  ... -1.060778e+00             1.0   
...        ...       ...       ...  ...           ...             ...   
599   2.057221  0.152729 -1.157955  ...  5.219412e-01             0.0   
3531 -0.960211  0.688169  0.050008  ... -1.060778e+00             1.0   
3944 -1.029316  0.424286  1.364123  ...  1.049514e+00             0.0   
1836 -0.140202 -1.936384  0.205555  ... -5.631988e-03             1.0   
4090 -0.018251 -0.707782 -0.489928  ... -1.060778e+00             1.0   

      machine_type_B  machine_type_C  shift_day  shift_evening  shift_night  \
1152             1.0             0.0        0.0            0.0          1.0   
4473             0.0             0.0        1.0            0.0          0.0   
3679             0.0             0.0        1.0            0.0          0.0   
3186             0.0             0.0        0.0            0.0          1.0   
3826             0.0             0.0        0.0            0.0          0.0   
...              ...             ...        ...            ...          ...   
599              0.0             0.0        0.0            1.0          0.0   
3531             0.0             0.0        0.0            1.0          0.0   
3944             1.0             0.0        0.0            1.0          0.0   
1836             0.0             0.0        0.0            1.0          0.0   
4090             0.0             0.0        0.0            1.0          0.0   

      plant_region_Centro  plant_region_Norte  plant_region_Sur  
1152                  1.0                 0.0               0.0  
4473                  0.0                 1.0               0.0  
3679                  1.0                 0.0               0.0  
3186                  1.0                 0.0               0.0  
3826                  0.0                 1.0               0.0  
...                   ...                 ...               ...  
599                   0.0                 1.0               0.0  
3531                  1.0                 0.0               0.0  
3944                  0.0                 1.0               0.0  
1836                  1.0                 0.0               0.0  
4090                  0.0                 1.0               0.0  

[1122 rows x 38 columns]

In [30]:
X_train_clean

s1        s2        s3        s4        s5        s6        s7  \
744   0.043449 -1.607153  0.760567 -0.935112  0.668751 -1.338896  0.136881   
3295 -0.371532 -1.187802 -0.373986  1.092292 -0.410907  0.836689  0.045863   
3851  0.498209  0.065696 -0.888494  0.533978  0.907719 -1.107246  1.257276   
3140  0.000927 -1.155569  0.494467  0.723594 -0.062235  0.829963  0.042125   
5359 -1.910834 -0.039246 -0.709817 -0.551048 -1.754210 -0.087337  2.502567   
...        ...       ...       ...       ...       ...       ...       ...   
4180  2.091494  0.435898 -0.542545  1.311800 -0.984048  0.022935 -0.488554   
5753 -0.709986 -1.871540 -0.004507  1.528286 -0.155287  0.849734 -0.727795   
5794  0.496300 -1.563301 -0.443031 -0.346473 -0.163508 -0.298600 -0.073057   
5982  0.772040  0.539741 -0.622003 -0.468171  0.832753  1.343317 -0.341340   
971   1.524221 -0.206996  0.802687 -0.139225  1.031348 -0.326926  0.753638   

            s8        s9       s10  ...    dia_semana  machine_type_A  \
744  -0.684062  1.766819 -0.743932  ... -1.060778e+00             0.0   
3295  0.611010  1.698236  0.887536  ... -1.588352e+00             0.0   
3851 -1.001514  0.131629 -0.410719  ... -5.631988e-03             0.0   
3140 -2.443358 -1.072561 -0.100504  ...  1.049514e+00             0.0   
5359 -0.018251 -0.879842  0.034243  ...  1.049514e+00             0.0   
...        ...       ...       ...  ...           ...             ...   
4180 -0.222144  1.322044  1.178943  ... -2.342896e-16             1.0   
5753 -0.202989 -2.172811 -1.863407  ... -1.588352e+00             1.0   
5794  0.633268  0.002767  0.658037  ...  1.577088e+00             0.0   
5982 -0.865552 -0.359048  0.013273  ...  1.049514e+00             0.0   
971   1.967889  0.392371 -0.890939  ...  5.219412e-01             0.0   

      machine_type_B  machine_type_C  shift_day  shift_evening  shift_night  \
744              0.0             0.0        1.0            0.0          0.0   
3295             0.0             1.0        0.0            1.0          0.0   
3851             1.0             0.0        1.0            0.0          0.0   
3140             0.0             1.0        0.0            0.0          1.0   
5359             0.0             1.0        1.0            0.0          0.0   
...              ...             ...        ...            ...          ...   
4180             0.0             0.0        1.0            0.0          0.0   
5753             0.0             0.0        1.0            0.0          0.0   
5794             1.0             0.0        0.0            0.0          1.0   
5982             0.0             1.0        1.0            0.0          0.0   
971              1.0             0.0        0.0            0.0          0.0   

      plant_region_Centro  plant_region_Norte  plant_region_Sur  
744                   1.0                 0.0               0.0  
3295                  1.0                 0.0               0.0  
3851                  1.0                 0.0               0.0  
3140                  1.0                 0.0               0.0  
5359                  1.0                 0.0               0.0  
...                   ...                 ...               ...  
4180                  0.0                 0.0               1.0  
5753                  1.0                 0.0               0.0  
5794                  0.0                 1.0               0.0  
5982                  1.0                 0.0               0.0  
971                   1.0                 0.0               0.0  

[4488 rows x 38 columns]

---

## 3. Modelado
- Entrenar al menos un modelo individual (baseline).
- Entrenar un modelo de ensamble usando bagging, boosting, voting o stacking.

In [31]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout

In [32]:
y_train.sort_values(), y_test.sort_values()

(2165    -964.369440
 3637    -872.387102
 5533    -612.386592
 3724    -497.004781
 5299    -467.517203
            ...     
 5942    1221.521707
 3287    1221.815071
 4950    1223.599796
 5465    1757.873710
 5499    1942.546474
 Name: quality_score, Length: 4488, dtype: float64,
 48      -729.685242
 3720    -599.287801
 1235     -17.503166
 784       -5.292434
 4716      -5.155232
            ...     
 425      958.231097
 2199    1087.368577
 5491    1094.294042
 3554    1162.416908
 2382    1511.326092
 Name: quality_score, Length: 1122, dtype: float64)

### Modelo Baseline

In [33]:
INPUT_SHAPE = (38,) 

model_baseline = Sequential([
    Dense(64, activation='relu', input_shape=INPUT_SHAPE),

    Dense(32, activation='relu'),
    
    Dense(1) 
])

model_baseline.compile(optimizer='adam', loss='mean_squared_error')

print(model_baseline.summary())

print("Entrenando modelo Baseline (Keras)...")
history = model_baseline.fit(
    X_train_clean, 
    y_train,
    epochs=50,
    batch_size=32,
    validation_data=(X_test_clean, y_test)
)

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense (Dense)                        │ (None, 64)                  │           2,496 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 32)                  │           2,080 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 1)                   │              33 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 4,609 (18.00 KB)

 Trainable params: 4,609 (18.00 KB)

 Non-trainable params: 0 (0.00 B)

None
Entrenando modelo Baseline (Keras)...
Epoch 1/50
141/141 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - loss: 15912.5752 - val_loss: 13929.2471
Epoch 2/50
141/141 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 9702.0186 - val_loss: 10980.2793
Epoch 3/50
141/141 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 8581.6689 - val_loss: 10276.6279
Epoch 4/50
141/141 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 6653.1753 - val_loss: 9743.0781
Epoch 5/50
141/141 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 10032.5195 - val_loss: 9192.9795
Epoch 6/50
141/141 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 7286.3809 - val_loss: 8720.6826
Epoch 7/50
141/141 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 7213.6313 - val_loss: 8320.1885
Epoch 8/50
141/141 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 5576.1504 - val_loss: 8012.0581
Epoch 9/50
141/141 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 5581.6138 - val_loss: 7702.4775
Epoch 10/50
141/141 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 7049.5293 - val_loss: 7431.9736
Epoch 11/50
141/141 ━━━━━━━━━━━

### Bagging

In [36]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error # Los necesitarás para el paso 4

rf_model = RandomForestRegressor(n_estimators=100, 
                                 random_state=42, 
                                 n_jobs=-1)

print("Entrenando modelo de Ensamble (Random Forest)...")
rf_model.fit(X_train_clean, y_train)

Entrenando modelo de Ensamble (Random Forest)...


RandomForestRegressor(n_jobs=-1, random_state=42)

In [37]:
from sklearn.metrics import mean_squared_error, mean_absolute_error

print("Evaluando Random Forest...")
y_pred_rf = rf_model.predict(X_test_clean)

mse_rf = mean_squared_error(y_test, y_pred_rf)

rmse_rf = np.sqrt(mse_rf)

mae_rf = mean_absolute_error(y_test, y_pred_rf)

print(f"Random Forest - RMSE: {rmse_rf:.2f}")
print(f"Random Forest - MAE: {mae_rf:.2f}")

Evaluando Random Forest...
Random Forest - RMSE: 75.93
Random Forest - MAE: 19.23


---
## 4. Evaluación
Aplicar métricas adecuadas al tipo de problema (ej.: accuracy, precision, recall, F1-score, ROC-AUC en clasificación; RMSE, MAE en regresión).